### PR TITLE
Weak segments performance

### PR DESCRIPTION
**Pull request by Nadav-Barak**
Resolves https://github.com/deepchecks/deepchecks/issues/1571

This is a big issue and so we will be uploaded and reviewed in parts for your connivance.

### CODE REVIEW

File: deepchecks/utils/performance/partition.py [hunk 1]
```diff
@@ -22,27 +23,38 @@
...
+        else:
+            self.filter_func = filter_functions
         self.label = label
 
-    def filter(self, dataframe):
-        """Run the filter on given dataframe."""
-        return dataframe.loc[self.filter_func(dataframe)]
+    def filter(self, dataframe: pd.DataFrame) -> pd.DataFrame:
+        """Run the filter on given dataframe. Return rows in data frame satisfying the filter properties."""
+        data = dataframe.copy()
```
Comment (matanper) [inline deepchecks/utils/performance/partition.py:49]: why copy?

File: deepchecks/utils/performance/partition.py [hunk 2]
```diff
@@ -142,11 +154,41 @@ def partition_column(
...
+    Returns
+    -------
+    List[DeepchecksFilter]:
+           A list of filters describing the leaves
+    """
+    node_to_feature = [feature_names[i] if i != _tree.TREE_UNDEFINED else None for i in tree.feature]
+
+    def recurse(node_idx, filter_up_to_node):
+        if tree.feature[node_idx] != _tree.TREE_UNDEFINED:
+            left_filter = DeepchecksFilter([lambda df, a=tree.threshold[node_idx]: df[node_to_feature[node_idx]] <= a])
```
Comment (matanper) [inline deepchecks/utils/performance/partition.py:186]: `node_to_feature` might contain Nones won't this fail `node_to_feature[node_idx]] <= a`?

Comment (matanper) [inline deepchecks/utils/performance/partition.py:186]: We use the filter like `dc_filter.filter(df)` where is the `a` parameter coming from?

File: deepchecks/utils/performance/partition.py [hunk 3]
```diff
@@ -142,11 +154,41 @@ def partition_column(
...
+           was trained on.
+
+    Returns
+    -------
+    List[DeepchecksFilter]:
+           A list of filters describing the leaves
+    """
+    node_to_feature = [feature_names[i] if i != _tree.TREE_UNDEFINED else None for i in tree.feature]
+
+    def recurse(node_idx, filter_up_to_node):
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/utils/performance/partition.py:184]: No, its truncate (intersect) filters and in that way returns the "complex" filters equivalent to the leaves of the tree

File: deepchecks/utils/performance/partition.py (hunk 1 above)
Comment (Nadav-Barak) [reply deepchecks/utils/performance/partition.py:49]: You are correct

File: deepchecks/utils/performance/partition.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/utils/performance/partition.py:184]: as a list - each element in the list correspond to a leaf

[the remark this answers is not in the document]
Comment (matanper) [reply deepchecks/utils/performance/partition.py:184]: so the usage of `DeepchecksFilter` is confusing, I'd expect the function to return list of `DeepchecksFilter`s and not a single one

File: deepchecks/utils/performance/partition.py [hunk 4]
```diff
@@ -22,27 +23,38 @@
...
 class DeepchecksFilter:
     """Contains a filter function which works on a dataframe and a label describing the filter.
 
     Parameters
     ----------
-    filter_func : Callable
-        function which receive dataframe and return a filter on it
-    label : str
+    filter_functions : Callable, default: None
+        List of functions which receive dataframe and return a filter on it. If None uses applies no filter
```
Comment (nirhutnik) [inline deepchecks/utils/performance/partition.py:35]:
```suggestion
        List of functions that receive a DataFrame and return a filter on it. If None, no filter is applied
```

File: deepchecks/utils/performance/partition.py [hunk 5]
```diff
@@ -22,27 +23,38 @@
...
+    label : str, default = ''
         name of the filter
     """
 
-    def __init__(self, filter_func: Callable, label: str):
-        self.filter_func = filter_func
+    def __init__(self, filter_functions: List[Callable] = None, label: str = ''):
+        if not filter_functions or len(filter_functions) == 0:
```
Comment (nirhutnik) [inline deepchecks/utils/performance/partition.py:41]: if len(filter_functions) == 0 is True, then "not filter_functions" will be True as well so redundant

File: deepchecks/utils/performance/partition.py [hunk 6]
```diff
@@ -22,27 +23,38 @@
...
 
 class DeepchecksFilter:
     """Contains a filter function which works on a dataframe and a label describing the filter.
 
     Parameters
     ----------
-    filter_func : Callable
-        function which receive dataframe and return a filter on it
-    label : str
+    filter_functions : Callable, default: None
```
Comment (nirhutnik) [inline deepchecks/utils/performance/partition.py:34]:
```suggestion
    filter_functions : List[Callable], default: None
```

File: deepchecks/utils/performance/partition.py [hunk 7]
```diff
@@ -22,27 +23,38 @@
...
+        List of functions which receive dataframe and return a filter on it. If None uses applies no filter
+    label : str, default = ''
```
Comment (nirhutnik) [inline deepchecks/utils/performance/partition.py:36]: Maybe have another default, like "filter" or something? Not critical.

File: deepchecks/utils/performance/partition.py [hunk 8]
```diff
@@ -22,27 +23,38 @@
...
         name of the filter
     """
 
-    def __init__(self, filter_func: Callable, label: str):
-        self.filter_func = filter_func
+    def __init__(self, filter_functions: List[Callable] = None, label: str = ''):
+        if not filter_functions or len(filter_functions) == 0:
+            self.filter_func = [lambda df: [True] * len(df)]
+        else:
+            self.filter_func = filter_functions
```
Comment (nirhutnik) [inline deepchecks/utils/performance/partition.py:44]: Should be called "filter_functions" in self as well IMO

File: deepchecks/utils/performance/partition.py [hunk 9]
```diff
@@ -142,11 +154,41 @@ def partition_column(
...
+    feature_names : List[str]
+        The feature names for elements within the tree. Normally it is the column names of the data frame the tree
+           was trained on.
+
+    Returns
+    -------
+    List[DeepchecksFilter]:
+           A list of filters describing the leaves
+    """
+    node_to_feature = [feature_names[i] if i != _tree.TREE_UNDEFINED else None for i in tree.feature]
```
Comment (nirhutnik) [inline deepchecks/utils/performance/partition.py:182]: If i is not an integer, give it another name

File: deepchecks/utils/performance/partition.py [hunk 10]
```diff
@@ -142,11 +154,41 @@ def partition_column(
...
         if len(cat_hist_dict) > n_large_cats:
             f = lambda df, values=cat_hist_dict.index[:n_large_cats]: ~df[column_name].isin(values)
-            filters.append(DeepchecksFilter(f, 'Others'))
+            filters.append(DeepchecksFilter([f], 'Others'))
 
         return filters
+
+
+def convert_tree_leaves_into_filters(tree, feature_names: List[str]) -> List[DeepchecksFilter]:
+    """Extract the leaves from a sklearn tree and covert them into DeepchecksFilters.
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/utils/performance/partition.py:168]: In general - really awesome, but you need more documentation and explanations on this function so in the future, others will get what it's doing :)

**Review by nirhutnik — CHANGES_REQUESTED** [review]

**Review by matanper — APPROVED** [review]

File: deepchecks/utils/performance/partition.py [hunk 11]
```diff
@@ -142,11 +153,45 @@ def partition_column(
...
+
+    More specifically go over the tree from root to leaf and truncates (intersect) the relevant filters along the way.
```
Comment (nirhutnik) [inline deepchecks/utils/performance/partition.py:169]:
```suggestion
    The function goes over the tree from root to leaf and concatenates (by intersecting) the relevant filters along the way.
```

File: deepchecks/utils/performance/partition.py [hunk 12]
```diff
@@ -142,11 +153,45 @@ def partition_column(
...
+    The result is a list in which each element is the truncate path between the root to a different leaf.
```
Comment (nirhutnik) [inline deepchecks/utils/performance/partition.py:170]:
```suggestion
    The function returns a list in which each element is a DeepchecksFilter representing the path between the root to a different leaf.
```

**Pull request merged**